aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -143,34 +147,85 @@ static void check_1d(const Tensor& t, const char* arg, const char* fn) {
+                              const Scalar scalar,
+                              const std::string& scalar_name) {
+  TORCH_CHECK(
+    !scalar.isBoolean() || dtype == ScalarType::Bool,
+    "Boolean ", scalar_name, " only supported for Boolean results.");
+  TORCH_CHECK(
+    isFloatingType(dtype) || isComplexType(dtype) || scalar.isIntegral(true),
+    "For integral input tensors, "
+    "argument ", scalar_name ," must not be a floating point number.");
+}
+
+static TensorIterator build_addr_iter(Tensor& result,
+                                      const Tensor& self,
+                                      const Tensor& vec1,
+                                      const Tensor& vec2) {
+  check_1d(vec1, "vec1", "addr");
+  check_1d(vec2, "vec2", "addr");
 
-  Tensor outer_result = at::outer(vec1, vec2) * alpha;
-  if (beta.to<double>() == 0.0) {
-    return outer_result;
+  Tensor self_;
+  if (&result != &self) {
+    std::tie(self_) = expand_size(self, {vec1.size(0), vec2.size(0)}, "addr");
+  } else {
+    self_ = self;
   }
-  return outer_result + (self * beta);
+  TORCH_CHECK(
+    self_.dim() == 2,
+    "2D tensor expected, got ", self_.dim(), "D tensor for input"
+  );
+  TORCH_CHECK(
+    self_.size(0) == vec1.size(0) && self_.size(1) == vec2.size(0),
+    "size mismatch, input: ", self_.sizes(),
+    ", v1: ", vec1.sizes(),
+    ", v2: ", vec2.sizes()
+  );
+
+  auto iter = TensorIteratorConfig()
+    .set_check_mem_overlap(true)
+    .add_output(result)
+    .add_input(self_)
+    .add_input(vec1.reshape({vec1.size(0), 1}))
+    .add_input(vec2)
+    .allow_cpu_scalars(true)
+    .promote_inputs_to_common_dtype(true)
+    .cast_common_dtype_to_outputs(true)
+    .enforce_safe_casting_to_output(true)
+    .build();
+  return iter;
+}
+
+Tensor addr(const Tensor& self,
+            const Tensor& vec1, const Tensor& vec2,
+            Scalar beta, Scalar alpha) {
+  Tensor result;
+  auto iter = build_addr_iter(result, self, vec1, vec2);
+
+  check_addr_scalar(iter.dtype(), beta, "beta");
+  check_addr_scalar(iter.dtype(), alpha, "alpha");
+
+  addr_stub(iter.device_type(), iter, beta, alpha);
+  return iter.output();
 }
 
-Tensor& addr_(Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
+Tensor& addr_(Tensor& self,
+              const Tensor& vec1, const Tensor& vec2,
+              Scalar beta, Scalar alpha) {
   return at::addr_out(self, self, vec1, vec2, beta, alpha);
 }
 
-Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
-  auto addr_result = at::addr(self, vec1, vec2, beta, alpha);
-  // Validates safe casting
-  const auto result_dtype = addr_result.scalar_type();
-  TORCH_CHECK(canCast(result_dtype, result.scalar_type()),
-              "result type ", result_dtype,
-              " can't be cast to the desired output type ", result.scalar_type());
+Tensor& addr_out(Tensor &result,
```

Comment:
One difference made in this PR is the old implementation works for all backends but now it only support CPU and CUDA only. 
So here are a few missing changes in this PR:
1. change the addr.out entry in native_functions.yaml from DefaultBackend to CPU,CUDA only.  
2. Save the old implementation of addr_out in a math_addr_out function and put it as Math keyword in native_functions.yaml. see https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#choosing-the-right-dispatch-keyword for more details. 
3. Add a test in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/math_kernel_test.cpp to make sure the new kernels you wrote for CPU & CUDA produces the same results as the old kernels we have. 
The second item should fix the xla test failure properly. 
All of the steps above should be smalll changes I think. 
Please let me know if you have followup questions! Thanks!

Reply:
@ailzhang Thank you so much for your kind help in fixing the XLA failures, and providing the detailed docs and samples.
I have addressed the suggested changes and the XLA CI checks passed.